File: reference/python/perception.py
"""Perspicacity v1 — Pure Python dataclasses for scene perception.

Zero dependencies. No bpy, no numpy.
"""
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Dict, List, Optional


@dataclass
class Vec3:
    x: float = 0.0
    y: float = 0.0
    z: float = 0.0


# ---------------------------------------------------------------------------
# Tier 0 — Identity
# ---------------------------------------------------------------------------

@dataclass
class SceneHeader:
    obj_count: int = 0
    light_count: int = 0
    energy: float = 0.0
    engine: str = ""
    ground_z: Optional[float] = None


@dataclass
class CameraData:
    name: str = ""
    position: Vec3 = field(default_factory=Vec3)
    focal_mm: float = 50.0


@dataclass
class LightData:
    name: str = ""
    type: str = "POINT"  # POINT | SUN | SPOT | AREA
    energy_w: float = 0.0
    color: Vec3 = field(default_factory=Vec3)
    position: Vec3 = field(default_factory=Vec3)


@dataclass
class ObjectData:
    name: str = ""
    position: Vec3 = field(default_factory=Vec3)
    coverage: float = 0.0
    quadrant: str = "mid-center"
    depth: float = 0.0
    material: str = ""
    dimensions: Optional[Vec3] = None
    top_z: Optional[float] = None  # AABB max Z — surface height for placement
    source: Optional[str] = None  # semantic name for hierarchy children
    rotation: Optional[Vec3] = None  # euler degrees, only when any component > 1°
    facing: Optional[str] = None  # compass: N/NE/E/SE/S/SW/W/NW
    zone: Optional[str] = None  # user-defined spatial zone
    face: Optional[str] = None
    lum: Optional[float] = None
    transparent: bool = False
    has_uv: Optional[bool] = None
    flipped_normals_pct: Optional[float] = None
    non_manifold_edges: Optional[int] = None
    inside: Optional[str] = None
    contains: List[str] = field(default_factory=list)


@dataclass
class SemanticGroupData:
    name: str = ""
    position: Vec3 = field(default_factory=Vec3)
    dimensions: Vec3 = field(default_factory=Vec3)
    top_z: float = 0.0
    material: str = ""
    facing: Optional[str] = None
    member_count: int = 0


@dataclass
class AssemblyData:
    name: str = ""
    members: List[str] = field(default_factory=list)
    center: Vec3 = field(default_factory=Vec3)
    types: str = ""


# ---------------------------------------------------------------------------
# Tier 1 — Spatial
# ---------------------------------------------------------------------------

@dataclass
class RelationshipData:
    from_obj: str = ""
    to_obj: str = ""
    distance: float = 0.0
    direction: str = ""
    vertical: Optional[str] = None
    overlap: bool = False
    overlap_pct: Optional[float] = None
    aabb_overlap_pct: Optional[float] = None
    contact: bool = False
    occludes: bool = False
    occ_pct: Optional[float] = None


@dataclass
class VerifyResult:
    """Post-transform verification — only emitted on failure."""
    object: str = ""
    message: str = ""


@dataclass
class ContainmentData:
    outer: str = ""
    inner: str = ""
    mode: str = "full"  # full | partial


@dataclass
class SpatialFact:
    object: str = ""
    type: str = ""  # bbox_below_surface, bbox_extends_into, scale_diagonal, etc.
    details: Dict[str, object] = field(default_factory=dict)


# ---------------------------------------------------------------------------
# Tier 2 — Visual
# ---------------------------------------------------------------------------

@dataclass
class CompositionData:
    thirds: float = 0.0
    visible: int = 0
    total: int = 0
    balance: float = 0.0
    depth: str = "1/3"
    edge: List[str] = field(default_factory=list)


@dataclass
class RayGridData:
    resolution: int = 0
    coverage: Dict[str, float] = field(default_factory=dict)
    empty: float = 0.0


@dataclass
class MultiViewData:
    view: str = ""
    coverage: Dict[str, float] = field(default_factory=dict)


# ---------------------------------------------------------------------------
# Tier 3 — Physical
# ---------------------------------------------------------------------------

@dataclass
class LightAnalysisData:
    light: str = ""
    surface: str = ""
    angle: float = 0.0
    intensity: float = 0.0
    effective: Optional[float] = None
    raw_intensity: Optional[float] = None
    shadow: List[str] = field(default_factory=list)


@dataclass
class ShadowAnalysisData:
    light: str = ""
    surface: str = ""
    coverage: float = 0.0
    casters: List[str] = field(default_factory=list)
    contact: bool = False
    gap: Optional[float] = None


@dataclass
class MaterialPrediction:
    name: str = ""
    appearance: str = ""
    needs: Optional[str] = None
    warning: Optional[str] = None


@dataclass
class HarmonyData:
    types: str = ""
    temperature: str = ""


@dataclass
class PaletteData:
    luminance: Optional[float] = None
    palette: List[str] = field(default_factory=list)


@dataclass
class WorldData:
    bg_color: Optional[Vec3] = None
    hdri: bool = False
    strength: float = 1.0


@dataclass
class PhysicsState:
    name: str = ""
    type: str = "static"  # static | dynamic | kinematic
    mass_kg: float = 0.0
    velocity: Optional[Vec3] = None
    sleeping: bool = False


@dataclass
class ContactData:
    obj_a: str = ""
    obj_b: str = ""
    normal: Vec3 = field(default_factory=Vec3)
    force_n: float = 0.0
    surface: str = ""


@dataclass
class SoundData:
    source: str = ""
    type: str = "point"  # ambient | point | directional
    volume: float = 0.0
    distance: float = 0.0
    direction: str = ""
    occlusion: float = 0.0


# ---------------------------------------------------------------------------
# Tier 4 — Semantic
# ---------------------------------------------------------------------------

@dataclass
class HierarchyEntry:
    chain: List[str] = field(default_factory=list)


@dataclass
class GroupEntry:
    name: str = ""
    members: List[str] = field(default_factory=list)


# ---------------------------------------------------------------------------
# Tier 5 — Temporal
# ---------------------------------------------------------------------------

@dataclass
class AnimationState:
    name: str = ""
    action: str = ""
    frame: int = 0
    total: int = 0
    playing: bool = False


# ---------------------------------------------------------------------------
# Tier containers
# ---------------------------------------------------------------------------

@dataclass
class IdentityTier:
    scene_header: Optional[SceneHeader] = None
    cameras: List[CameraData] = field(default_factory=list)
    lights: List[LightData] = field(default_factory=list)
    objects: List[ObjectData] = field(default_factory=list)
    semantic_groups: List[SemanticGroupData] = field(default_factory=list)
    assemblies: List[AssemblyData] = field(default_factory=list)
    focus: Optional['FocusData'] = None


@dataclass
class SpatialTier:
    relationships: List[RelationshipData] = field(default_factory=list)
    verify: List[VerifyResult] = field(default_factory=list)
    containment: List[ContainmentData] = field(default_factory=list)
    spatial_facts: List[SpatialFact] = field(default_factory=list)


@dataclass
class VisualTier:
    composition: Optional[CompositionData] = None
    ray_grid: Optional[RayGridData] = None
    multi_views: List[MultiViewData] = field(default_factory=list)


@dataclass
class PhysicalTier:
    light_analyses: List[LightAnalysisData] = field(default_factory=list)
    shadow_analyses: List[ShadowAnalysisData] = field(default_factory=list)
    materials: List[MaterialPrediction] = field(default_factory=list)
    harmony: Optional[HarmonyData] = None
    palette: Optional[PaletteData] = None
    world: Optional[WorldData] = None
    physics_states: List[PhysicsState] = field(default_factory=list)
    contacts: List[ContactData] = field(default_factory=list)
    sounds: List[SoundData] = field(default_factory=list)


@dataclass
class SemanticTier:
    hierarchy: List[HierarchyEntry] = field(default_factory=list)
    groups: List[GroupEntry] = field(default_factory=list)


@dataclass
class TemporalTier:
    animations: List[AnimationState] = field(default_factory=list)
    deltas: List[str] = field(default_factory=list)


@dataclass
class FocusData:
    """Proximity-based perception query parameters."""
    position: Vec3 = field(default_factory=Vec3)
    radius: float = 0.0
    near: int = 0
    mid: int = 0
    far: int = 0
    out: int = 0


# ---------------------------------------------------------------------------
# Top-level struct
# ---------------------------------------------------------------------------

@dataclass
class ScenePerception:
    focus: Optional[FocusData] = None
    viewpoint: str = "Camera"
    viewpoint_position: Vec3 = field(default_factory=Vec3)
    viewpoint_forward: Vec3 = field(default_factory=lambda: Vec3(0, 0, -1))
    identity: IdentityTier = field(default_factory=IdentityTier)
    spatial: SpatialTier = field(default_factory=SpatialTier)
    visual: VisualTier = field(default_factory=VisualTier)
    physical: PhysicalTier = field(default_factory=PhysicalTier)
    semantic: SemanticTier = field(default_factory=SemanticTier)
    temporal: TemporalTier = field(default_factory=TemporalTier)

File: reference/zig/perception.zig
//! Perspicacity v1 -- Pure Zig structs for scene perception.
//!
//! Zero dependencies beyond std. All heap allocations go through
//! the caller-provided `std.mem.Allocator`.

const std = @import("std");
const Allocator = std.mem.Allocator;

// ---------------------------------------------------------------------------
// Vec3
// ---------------------------------------------------------------------------

pub const Vec3 = struct {
    x: f32 = 0.0,
    y: f32 = 0.0,
    z: f32 = 0.0,
};

// ---------------------------------------------------------------------------
// Tier 0 -- Identity
// ---------------------------------------------------------------------------

pub const SceneHeader = struct {
    obj_count: i32 = 0,
    light_count: i32 = 0,
    energy: f32 = 0.0,
    engine: []const u8 = "",
    ground_z: ?f32 = null,
};

pub const CameraData = struct {
    name: []const u8 = "",
    position: Vec3 = .{},
    focal_mm: f32 = 50.0,
};

pub const LightData = struct {
    name: []const u8 = "",
    light_type: []const u8 = "POINT", // POINT | SUN | SPOT | AREA
    energy_w: f32 = 0.0,
    color: Vec3 = .{},
    position: Vec3 = .{},
};

pub const ObjectData = struct {
    name: []const u8 = "",
    position: Vec3 = .{},
    coverage: f32 = 0.0,
    quadrant: []const u8 = "mid-center",
    depth: f32 = 0.0,
    material: []const u8 = "",
    dimensions: ?Vec3 = null,
    top_z: ?f32 = null,
    rotation: ?Vec3 = null,
    facing: ?[]const u8 = null,
    zone: ?[]const u8 = null,
    face: ?[]const u8 = null,
    lum: ?f32 = null,
    transparent: bool = false,
    has_uv: ?bool = null,
    inside: ?[]const u8 = null,
    contains: []const []const u8 = &.{},
};

pub const SemanticGroupData = struct {
    name: []const u8 = "",
    position: Vec3 = .{},
    dimensions: Vec3 = .{},
    top_z: f32 = 0.0,
    material: []const u8 = "",
    facing: ?[]const u8 = null,
    member_count: i32 = 0,
};

pub const AssemblyData = struct {
    name: []const u8 = "",
    members: []const []const u8 = &.{},
    center: Vec3 = .{},
    types: []const u8 = "",
};

// ---------------------------------------------------------------------------
// Tier 1 -- Spatial
// ---------------------------------------------------------------------------

pub const RelationshipData = struct {
    from_obj: []const u8 = "",
    to_obj: []const u8 = "",
    distance: f32 = 0.0,
    direction: []const u8 = "",
    vertical: ?[]const u8 = null,
    overlap: bool = false,
    overlap_pct: ?f32 = null,
    aabb_overlap_pct: ?f32 = null,
    contact: bool = false,
    occludes: bool = false,
    occ_pct: ?f32 = null,
};

pub const VerifyResult = struct {
    object: []const u8 = "",
    message: []const u8 = "",
};

pub const ContainmentData = struct {
    outer: []const u8 = "",
    inner: []const u8 = "",
    mode: []const u8 = "full",
};

pub const SpatialFact = struct {
    object: []const u8 = "",
    fact_type: []const u8 = "",
    // Details stored as parallel key/value slices
    detail_keys: []const []const u8 = &.{},
    detail_vals: []const []const u8 = &.{},
};

// ---------------------------------------------------------------------------
// Tier 2 -- Visual
// ---------------------------------------------------------------------------

pub const CompositionData = struct {
    thirds: f32 = 0.0,
    visible: i32 = 0,
    total: i32 = 0,
    balance: f32 = 0.0,
    depth: []const u8 = "1/3",
    edge: []const []const u8 = &.{},
};

pub const RayGridData = struct {
    resolution: i32 = 0,
    coverage_keys: []const []const u8 = &.{},
    coverage_vals: []const f32 = &.{},
    empty: f32 = 0.0,

    /// Look up a coverage value by object name.
    pub fn getCoverage(self: *const RayGridData, key: []const u8) ?f32 {
        for (self.coverage_keys, 0..) |k, i| {
            if (std.mem.eql(u8, k, key)) return self.coverage_vals[i];
        }
        return null;
    }
};

pub const MultiViewData = struct {
    view: []const u8 = "",
    coverage_keys: []const []const u8 = &.{},
    coverage_vals: []const f32 = &.{},

    pub fn getCoverage(self: *const MultiViewData, key: []const u8) ?f32 {
        for (self.coverage_keys, 0..) |k, i| {
            if (std.mem.eql(u8, k, key)) return self.coverage_vals[i];
        }
        return null;
    }
};

// ---------------------------------------------------------------------------
// Tier 3 -- Physical
// ---------------------------------------------------------------------------

pub const LightAnalysisData = struct {
    light: []const u8 = "",
    surface: []const u8 = "",
    angle: f32 = 0.0,
    intensity: f32 = 0.0,
    shadow: []const []const u8 = &.{},
};

pub const ShadowAnalysisData = struct {
    light: []const u8 = "",
    surface: []const u8 = "",
    coverage: f32 = 0.0,
    casters: []const []const u8 = &.{},
    contact: bool = false,
    gap: ?f32 = null,
};

pub const MaterialPrediction = struct {
    name: []const u8 = "",
    appearance: []const u8 = "",
    needs: ?[]const u8 = null,
    warning: ?[]const u8 = null,
};

pub const HarmonyData = struct {
    types: []const u8 = "",
    temperature: []const u8 = "",
};

pub const PaletteData = struct {
    luminance: ?f32 = null,
    palette: []const []const u8 = &.{},
};

pub const WorldData = struct {
    bg_color: ?Vec3 = null,
    hdri: bool = false,
    strength: f32 = 1.0,
};

pub const PhysicsState = struct {
    name: []const u8 = "",
    phys_type: []const u8 = "static", // static | dynamic | kinematic
    mass_kg: f32 = 0.0,
    velocity: ?Vec3 = null,
    sleeping: bool = false,
};

pub const ContactData = struct {
    obj_a: []const u8 = "",
    obj_b: []const u8 = "",
    normal: Vec3 = .{},
    force_n: f32 = 0.0,
    surface: []const u8 = "",
};

pub const SoundData = struct {
    source: []const u8 = "",
    sound_type: []const u8 = "point", // ambient | point | directional
    volume: f32 = 0.0,
    distance: f32 = 0.0,
    direction: []const u8 = "",
    occlusion: f32 = 0.0,
};

// ---------------------------------------------------------------------------
// Tier 4 -- Semantic
// ---------------------------------------------------------------------------

pub const HierarchyEntry = struct {
    chain: []const []const u8 = &.{},
};

pub const GroupEntry = struct {
    name: []const u8 = "",
    members: []const []const u8 = &.{},
};

// ---------------------------------------------------------------------------
// Tier 5 -- Temporal
// ---------------------------------------------------------------------------

pub const AnimationState = struct {
    name: []const u8 = "",
    action: []const u8 = "",
    frame: i32 = 0,
    total: i32 = 0,
    playing: bool = false,
};

// ---------------------------------------------------------------------------
// Tier containers
// ---------------------------------------------------------------------------

pub const IdentityTier = struct {
    scene_header: ?SceneHeader = null,
    cameras: []const CameraData = &.{},
    lights: []const LightData = &.{},
    objects: []const ObjectData = &.{},
    semantic_groups: []const SemanticGroupData = &.{},
    assemblies: []const AssemblyData = &.{},
};

pub const SpatialTier = struct {
    relationships: []const RelationshipData = &.{},
    verify: []const VerifyResult = &.{},
    containment: []const ContainmentData = &.{},
    spatial_facts: []const SpatialFact = &.{},
};

pub const VisualTier = struct {
    composition: ?CompositionData = null,
    ray_grid: ?RayGridData = null,
    multi_views: []const MultiViewData = &.{},
};

pub const PhysicalTier = struct {
    light_analyses: []const LightAnalysisData = &.{},
    shadow_analyses: []const ShadowAnalysisData = &.{},
    materials: []const MaterialPrediction = &.{},
    harmony: ?HarmonyData = null,
    palette: ?PaletteData = null,
    world: ?WorldData = null,
    physics_states: []const PhysicsState = &.{},
    contacts: []const ContactData = &.{},
    sounds: []const SoundData = &.{},
};

pub const SemanticTier = struct {
    hierarchy: []const HierarchyEntry = &.{},
    groups: []const GroupEntry = &.{},
};

pub const TemporalTier = struct {
    animations: []const AnimationState = &.{},
    deltas: []const []const u8 = &.{},
};

// ---------------------------------------------------------------------------
// Top-level struct
// ---------------------------------------------------------------------------

pub const ScenePerception = struct {
    viewpoint: []const u8 = "Camera",
    viewpoint_position: Vec3 = .{},
    viewpoint_forward: Vec3 = .{ .x = 0, .y = 0, .z = -1 },
    identity: IdentityTier = .{},
    spatial: SpatialTier = .{},
    visual: VisualTier = .{},
    physical: PhysicalTier = .{},
    semantic: SemanticTier = .{},
    temporal: TemporalTier = .{},
};
